File: samples/perl/crosswalk.pl
#!/usr/bin/perl -w

## usage: perl crosswalk.pl -u your-umls-username -p your-umls-password -v source vocabulary -i source vocabulary identifier [ -r specify target vocabulary]
## Performs a crosswalk using the latest UMLS version and retrieves all codes that share a UMLS CUI with a particular code.
## Example: 
## perl crosswalk.pl -u username -p password -v HPO -i HP:0001947
## perl crosswalk.pl -u username -p password -v HPO -i HP:0001947 -r SNOMEDCT_US
## More information: https://documentation.uts.nlm.nih.gov/rest/source-asserted-identifiers/crosswalk/index.html

use lib "lib";
use strict;
use URI;
use Authentication::TicketClient;
use JSON;
use REST::Client;
use Data::Dumper;
use Getopt::Std;

## parse command line arguments
our ($opt_u,$opt_p,$opt_v,$opt_i, $opt_r);
getopt('upvir');
my $username = $opt_u || die "please provide username";
my $password = $opt_p || die "please provide password";
my $source   = $opt_v || die "Please provide a source vocabulary";
my $identifier = $opt_i || die "Please provide a source vocabulary identifier";

## Create a ticket granting ticket for the session
my $ticketClient = 
  new TicketClient(username=>$opt_u,password=>$opt_p,service=>"http://umlsks.nlm.nih.gov",tgt=>"") || die "could not create TicketClient() object";
my $tgt = $ticketClient->getTgt();
my $uri = new URI("https://uts-ws.nlm.nih.gov");
my $json;
my $client = REST::Client->new();
my %parameters = ();

my $path = sprintf("/rest/crosswalk/current/source/%s/%s", $opt_v, $opt_i);

if ( defined $opt_r && $opt_r ){
  $parameters{'targetSource'} = $opt_r
}

## Query the API
$json = run_query($path,\%parameters);
my $ra_results = $json->{'result'};

## Loop through results
foreach my $result( @$ra_results ) {
  printf("%s %s %s\n", $result->{'rootSource'}, $result->{'ui'}, $result->{'name'} )
}


sub format_json {
  my $json_in = shift;
  my $json    = JSON->new;
  my $obj     = $json->decode($json_in);
  return $obj;
}

sub run_query {
  my ($path, $parameters) = @_; 
  $parameters{ticket} = $ticketClient->getServiceTicket();
  $uri->path($path);
  $uri->query_form($parameters);
  print qq{$uri\n\n};
  my $query = $client->GET($uri) || die "Could not execute query $!\n";
  my $results = $query->responseCode() eq '200'? $query->responseContent: die "Could not execute query $!\n";
  my $json = format_json($results);
  return $json;
}

File: samples/python/crosswalk.py
## Steven Emrick - steve.emrick@nih.gov
## usage: python crosswalk.py -k <your-api-key>
## You can specify a specific UMLS version with the -v argument, but it is not required
## This reads a file with codes from the Human Phenotype Ontology and maps them to the US Edition of SNOMED CT through UMLS CUIs

from __future__ import print_function
from Authentication import *
import requests
import json
import argparse
import collections
import sys
import os

if sys.version_info < (3, 0):
    reload(sys)
    sys.setdefaultencoding('utf-8')

parser = argparse.ArgumentParser(description='process user given parameters')
parser.add_argument("-k", "--apikey", required = True, dest = "apikey", help = "enter api key from your UTS Profile")
parser.add_argument("-v", "--version", required =  False, dest="version", default = "current", help = "enter version example-2015AA")

args = parser.parse_args()
apikey = args.apikey
version = args.version
AuthClient = Authentication(apikey)

###################################
#get TGT for our session
###################################

tgt = AuthClient.gettgt()
base_uri = "https://uts-ws.nlm.nih.gov"
crosswalk_endpoint = "/rest/crosswalk/"+version+"/source/HPO"

def crosswalk_code(path):
    query = {'ticket': AuthClient.getst(tgt),'targetSource': 'SNOMEDCT_US'}
    r = requests.get(base_uri + path, params=query)
    #print(r.url + "\n")
    items = json.loads(r.text)
    return items

with open('hpo-codes.txt','r') as f:
     for line in f:
         ##get rid of newlines
         code = line.strip()
         path =  crosswalk_endpoint+"/"+code
         try:
             results = crosswalk_code(path)
             for sourceAtomCluster in results["result"]:
                 print('HPO Code - ' + code+ '\t' + 'SNOMEDCT concept -- ' + sourceAtomCluster["ui"] + ': ' + sourceAtomCluster["name"])

         except ValueError:
             print("No result found for "+code)
             pass

f.close()

